source/httpfs/driver.go

```diff
@@ -0,0 +1,159 @@
+package httpfs
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path"
+
+	"github.com/golang-migrate/migrate/v4/source"
+)
+
+// driver is a migration source driver for reading migrations from
+// http.FileSystem instances. It implements source.Driver interface and can be
+// used as a migration source for the main migrate library.
+type driver struct {
+	migrations *source.Migrations
+	fs         http.FileSystem
+	path       string
+}
+
+// New creates a new migrate source driver from a http.FileSystem instance and a
+// relative path to migration files within the virtual FS. It is identical to
+// the WithInstance() function except it will delay any errors on fist usage of
+// this driver. This reduces the number of error handling branches in clients of
+// this package without compromising on correctness.
+func New(fs http.FileSystem, path string) source.Driver {
```

Comment:
If we use the following `Init()` function signature:

```go
func (f *HTTPFS) Init(fs http.FileSystem, cfg Config) error
```

users of this package will have to use:

```go
	var src httpfs.HTTPFS
	if err := src.Init(http.Dir("sql"), httpfs.Config{}); err != nil {
		// do something
	}
	m, err := migrate.NewWithSourceInstance("httpfs", src, "database://url")
	if err != nil {
		// do something
	}
        err = m.Up()
	...
```

Not calling Init function would be a possible mistake that is not detectable by the compiler.

Could we use the following `Init()` function signature instead?

```go
func Init(fs http.FileSystem, cfg Config) (*HTTPFS, error)
```

Then the clients of this package would look like:

```go
	src, err := httpfs.Init(http.Dir("sql"), httpfs.Config{})
	if err != nil {
		// do something
	}
	m, err := migrate.NewWithSourceInstance("httpfs", src,	"database://url")
	if err != nil {
		// do something
	}
        err = m.Up()
	...
```

It would be impossible to create a new HTTPFS instance without calling `Init()`. Of course the name `Init()` would not be very accurate anymore. It is left here just to be consistent with the previous example.

What do you think?

Current implementation is focused around applications using this package directly. I understand your motivation to re-use this package for other source drivers as well. If we make this package easy to use directly all the other source drivers that wraps it will benefit from the ease of use as well.

Reply:
The current design using `New()` is awkward to use when developing a `source.Driver`.

The idea is is that the `httpfs` package would not be used directly. e.g. It'd always be "wrapped" as an embedded field in a struct which then acts as a `source.Driver`

Ideally, common `http.Filesystem`s would have a `source.Driver` implementation in migrate which reduces the work users of migrate would have to do. e.g. implementation and testing

Reply:
Would it be possible to leave two constructor function - `Init()` for embedding use-case and `New()` for direct access use-case?

Internally our use case for this package is being able to run migrations that are embedded into the main binary with github.com/shurcooL/vfsgen. This package works by creating a global variable that implements http.FileSystem interface. In this use case there is no need for wrapper driver that pulls http.FileSystem from somewhere deeper, we already have a variable that implements http.FileSystem and just need to convert it to a migrate driver. I think there could be a lot of similar use-cases when the application developer already have http.FileSystem or is able to acquire it in its own way. For such use-cases being able to conveniently convert http.FileSystem to migrate.Driver is really useful (including deferred error reporting).

Reply:
> Would it be possible to leave two constructor function - Init() for embedding use-case and New() for direct access use-case?
> ...
> we already have a variable that implements http.FileSystem and just need to convert it to a migrate driver. I think there could be a lot of similar use-cases when the application developer already have http.FileSystem or is able to acquire it in its own way

That sounds like a good idea!

Reply:
Actually I don't think we can have 2 "constructors" after removing the `Open()` method. You'll need a struct to wrap the `http.FileSystem`

Reply:
OK, latest version have two separate structs. One that is used for embedding into other drivers named `httpfs.Migrator` and one for converting `http.FileSystem` to `source.Driver` (not public).